src/Builder/Encoder/OperatorEncoder.php

```diff
@@ -0,0 +1,161 @@
+            // Skip optional arguments.
+            // $slice operator has the optional <position> argument in the middle of the array
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            $result[] = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    /**
+     * $group stage have a specific encoding because the _id argument is required and others are variadic
+     */
+    private function encodeAsGroup(GroupStage $value): stdClass
+    {
+        $result = new stdClass();
+        $result->_id = $this->recursiveEncode($value->_id);
+
+        foreach (get_object_vars($value->field) as $key => $val) {
+            $result->{$key} = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    private function encodeAsObject(OperatorInterface $value): stdClass
+    {
+        $result = new stdClass();
+        foreach (get_object_vars($value) as $key => $val) {
+            // Skip optional arguments. If they have a default value, it is resolved by the server.
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            $result->{$key} = $this->recursiveEncode($val);
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    private function encodeAsDollarObject(OperatorInterface $value): stdClass
+    {
+        $result = new stdClass();
+        foreach (get_object_vars($value) as $key => $val) {
+            // Skip optional arguments. If they have a default value, it is resolved by the server.
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            $val = $this->recursiveEncode($val);
+
+            if ($key === 'geometry') {
+                if (is_object($val) && property_exists($val, '$geometry')) {
+                    $result->{'$geometry'} = $val->{'$geometry'};
+                } elseif (is_array($val) && array_key_exists('$geometry', $val)) {
+                    $result->{'$geometry'} = $val['$geometry'];
+                } else {
+                    $result->{'$geometry'} = $val;
+                }
+            } else {
+                $result->{'$' . $key} = $val;
+            }
+        }
+
+        return $this->wrap($value, $result);
+    }
+
+    /**
+     * Get the unique property of the operator as value
+     */
+    private function encodeAsSingle(OperatorInterface $value): stdClass
+    {
+        foreach (get_object_vars($value) as $val) {
+            $result = $this->recursiveEncode($val);
+
+            return $this->wrap($value, $result);
+        }
+
+        throw new LogicException(sprintf('Class "%s" does not have a single property.', $value::class));
+    }
+
+    private function wrap(OperatorInterface $value, mixed $result): stdClass
+    {
+        $object = new stdClass();
+        $object->{$value->getOperator()} = $result;
+
+        return $object;
```

Comment:
Do we want a one-liner for this?

```suggestion
        return (object) [$value->getOperator() => $result];
```

Reply:
I'd like to microbench both implementations.

Reply:
Deferring to @GromNaN for this

Reply:
Since I was already microbenching `array_key_exists`, I also checked this. There's no significant performance difference (30 ms when testing 1M object creations), so there's no performance benefit in using either.